File: Web/Tools/Streams/AxMemoryStream.js
var StreamWriteMode = Object.freeze(
{
    Overwrite : 0,
    Insert : 1
});

/**
 * @constructor
 * Creates a memory stream object
 */
function AxMemoryStream()
{
    AxStream.call(this);

    this.capacity = 0;
    this.data = null;
}

AxMemoryStream.prototype = Object.create(AxStream.prototype);

AxMemoryStream.prototype.AxMemoryStream = function(initialCapacity)
{
    this.SetWriteMode(StreamWriteMode.Overwrite);
    
    this.capacity = 0;
    this.data = null;
    
    this.Resize(initialCapacity);
};

AxMemoryStream.prototype.AxMemoryStream_2 = function(data, size)
{
    this.length = size;

    this.SetWriteMode(StreamWriteMode.Overwrite);

    this.capacity = size;
    this.data = data;
};

/**
 * Reads data from the memory stream and returns the number of bytes which were successfully read
 * @param {ArrayBuffer} destination An array buffer of the appropriate size, into which the read data will be stored
 * @param {Integer} size Number of bytes to read
 * @return {Integer} The number of bytes which were successfully read
 */
AxMemoryStream.prototype.ReadData = function(destination, size)
{
    var bytesToCopy = AxMath.Min(this.length - this.position, size);
    AxMem.Copy(destination, 0, this.data, this.position, bytesToCopy);
    
    this.position += bytesToCopy;

    return bytesToCopy;
};

/**
 * Writes data to the memory stream and returns the number of bytes which were successfully written
 * If writing at a position before the end of the stream, depending on the stream's write mode, new data can either overwite bytes ahead of the position or be inserted - pushing the existing data forward.
 * @param {ArrayBuffer} data An array buffer which holds the data to be written
 * @param {Integer} size Number of bytes to write
 * @return {Integer} The number of bytes which were successfully written
 */
AxMemoryStream.prototype.WriteData = function(data, size)
{
    var requiredCapacity = this.writeMode === StreamWriteMode.Overwrite ? this.position + size : this.length + size;
    var newCapacity = this.capacity;
    while (requiredCapacity > newCapacity)
    {
        newCapacity *= 2;
    }
    if (newCapacity !== this.capacity)
        this.Resize(newCapacity);

    if (this.writeMode === StreamWriteMode.Insert)
    {
        var tailBytes = this.length - this.position;
        if (size > tailBytes)
        {
            AxMem.Copy(this.data, this.position + size, this.data, this.position, tailBytes);
        }
        else
        {
            var tailData = AxMem.Get(tailBytes);
            AxMem.Copy(tailData, 0, this.data, this.position, tailBytes);
            AxMem.Copy(this.data, this.position + size, tailData, 0, tailBytes);
        }

        this.length += size;
    }

    AxMem.Copy(this.data, this.position, data, 0, size);

    this.position += size;

    this.length = AxMath.Max(this.length, this.position);

    return size;
};

/**
 * Sets the write mode for the stream
 * If writing at a position before the end of the stream, depending on the stream's write mode, new data can either overwite bytes ahead of the position or be inserted - pushing the ahead bytes forward.
 * @param {StreamWriteMode} writeMode The write mode for the stream
 */
AxMemoryStream.prototype.SetWriteMode = function(writeMode)
{
    this.writeMode = writeMode;
};

/**
 * Resizes the memory stream to the given capacity in bytes. Existing data is saved as possible.
 * @param {Number} newCapacity The new capacity after the resize
 */
AxMemoryStream.prototype.Resize = function(newCapacity)
{
    var newData = AxMem.Get(newCapacity);

    if (this.length > 0)
        AxMem.Copy(newData, 0, this.data, 0, AxMath.Min(this.length, newCapacity));

    this.data = newData;
    this.capacity = newCapacity;
};

/**
 * Gets the ArrayBuffer which holds the data of the stream
 * @return {ArrayBuffer} The ArrayBuffer which holds the data of the stream
 */
AxMemoryStream.prototype.GetDataPointer = function()
{
	return this.data;
};

File: Engine/Axis/Source/Tools/Streams/AxMemoryStream.cpp
//---------------------------------------------------------------------------------------------
//	Axis 3D Engine Copyright (C) 2017 Palitri
//
//	This software is released under the MIT License http://palitri.com/axis/license/
//
//	For help and documentation, visit http://palitri.com
//---------------------------------------------------------------------------------------------

#include "AxMemoryStream.h"

#include "..\AxMem.h"
#include "..\AxMath.h"

AxMemoryStream::AxMemoryStream(int initialCapacity)
	: AxStream()
{
	this->SetWriteMode(StreamWriteMode_Overwrite);
	
	this->capacity = 0;
	this->data = 0;

	this->Resize(initialCapacity);
}

AxMemoryStream::AxMemoryStream(void *data, int size)
	: AxStream()
{
	this->length = size;

	this->SetWriteMode(StreamWriteMode_Overwrite);
	
	this->capacity = size;
	this->data = data;
}


AxMemoryStream::~AxMemoryStream(void)
{
	AxMem::FreeAndNull(&this->data);
}

long long AxMemoryStream::ReadData(void *data, long long size)
{
	long long bytesToCopy = AxMath::Min(this->length - this->position, size);
	AxMem::Copy(data, (void*)((unsigned int)this->data + (unsigned int)this->position), (unsigned int)bytesToCopy);

	this->position += bytesToCopy;

	return bytesToCopy;
}

long long AxMemoryStream::WriteData(const void *data, long long size)
{
	long long requiredCapacity = this->writeMode == StreamWriteMode_Overwrite ? this->position + size : this->length + size;
	long long newCapacity = this->capacity;
	while (requiredCapacity > newCapacity)
	{
		newCapacity *= 2;
	}
	if (newCapacity != this->capacity)
		this->Resize(newCapacity);

	if (this->writeMode == StreamWriteMode_Insert)
	{
		long long tailBytes = this->length - this->position;
		if (size > tailBytes)
		{
			AxMem::Copy(this->GetDataPointer((unsigned int)this->position + size), this->GetDataPointer(this->position), (unsigned int)tailBytes);
		}
		else
		{
			void *tailData = AxMem::Get((unsigned int)tailBytes);
			AxMem::Copy(tailData, this->GetDataPointer(this->position), (unsigned int)tailBytes);
			AxMem::Copy(this->GetDataPointer(this->position + size), tailData, (unsigned int)tailBytes);
			AxMem::FreeAndNull(&tailData);
		}

		this->length += size;
	}

	AxMem::Copy(this->GetDataPointer(this->position), data, (unsigned int)size);

	this->position += size;

	this->length = AxMath::Max(this->length, this->position);

	return size;
}

void AxMemoryStream::SetWriteMode(StreamWriteMode writeMode)
{
	this->writeMode = writeMode;
}

void AxMemoryStream::Resize(long long newCapacity)
{
	void *newData = AxMem::Get((int)newCapacity);
	
	if (this->length > 0)
		AxMem::Copy(newData, this->data, (unsigned int)AxMath::Min(this->length, newCapacity));
	if (this->data != 0)
		AxMem::FreeAndNull(&this->data);

	this->data = newData;
	this->capacity = newCapacity;

};

void *AxMemoryStream::GetDataPointer(long long offset)
{
	return (void*)((int)this->data + offset);
}
